File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums.length == 1) return nums[0];
        int res = nums[0]; 
        int currentMin = 1, currentMax = 1;
        for (int n : nums) {
            if (n == 0) {
                currentMin = 1; 
                currentMax = 1;
                res = Math.max(res, 0);
            } else {
                int temp = currentMax * n;
                currentMax = Math.max(Math.max(n, temp), currentMin * n);
                currentMin = Math.min(Math.min(n, temp), currentMin * n);
                res = Math.max(currentMax, res);
            }
        }
        return res;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 1 : return nums[0]
        res = nums[0]
        l = 1
        r = 1
        for i in range(len(nums)):
            l = 1 if l == 0 else l
            r = 1 if r == 0 else r
            l *= nums[i]
            r *= nums[len(nums)-1-i]
            res = max(res,l,r)
        return res
